library/src/main/java/com/novoda/downloadmanager/lib/CursorTranslator.java

```diff
@@ -0,0 +1,149 @@
+package com.novoda.downloadmanager.lib;
+
+import android.content.ContentUris;
+import android.database.Cursor;
+import android.database.CursorWrapper;
+import android.net.Uri;
+
+import java.io.File;
+
+/**
+ * This class wraps a cursor returned by DownloadProvider -- the "underlying cursor" -- and
+ * presents a different set of columns, those defined in the DownloadManager.COLUMN_* constants.
+ * Some columns correspond directly to underlying values while others are computed from
+ * underlying data.
+ */
+class CursorTranslator extends CursorWrapper {
+    private final Uri baseUri;
+    private final StatusTranslator statusTranslator;
+
+    public CursorTranslator(Cursor cursor, Uri baseUri, StatusTranslator statusTranslator) {
+        super(cursor);
+        this.baseUri = baseUri;
+        this.statusTranslator = statusTranslator;
+    }
+
+    @Override
+    public int getInt(int columnIndex) {
+        return (int) getLong(columnIndex);
+    }
+
+    @Override
+    public long getLong(int columnIndex) {
+        String columnName = getColumnName(columnIndex);
+        switch (columnName) {
+            case DownloadManager.COLUMN_REASON:
+                return getReason(super.getInt(getColumnIndex(DownloadContract.Downloads.COLUMN_STATUS)));
+            case DownloadManager.COLUMN_STATUS:
+                return getStatus(super.getInt(getColumnIndex(DownloadContract.Downloads.COLUMN_STATUS)));
+            case DownloadManager.COLUMN_BATCH_STATUS:
+                return getStatus(super.getInt(getColumnIndex(DownloadContract.Batches.COLUMN_STATUS)));
+            default:
+                return super.getLong(columnIndex);
+        }
+    }
+
+    private long getReason(int status) {
+        switch (getStatus(status)) {
```

Comment:
Would have been great to have a special treatment of error with status 418; `I'm a teapot`... actually NO, but the error is still funny. https://sitesdoneright.com/blog/2013/03/what-is-418-im-a-teapot-status-code-error